iothub/device/tests/Pipeline/ClientTransportStateMachineTests.cs

```diff
@@ -0,0 +1,138 @@
+        [DataRow(ClientTransportState.Open, ClientStateAction.ConnectionLost, ClientTransportState.Open)]
+        [DataRow(ClientTransportState.Open, ClientStateAction.ConnectionLost, ClientTransportState.Closing)]
+        [DataRow(ClientTransportState.Open, ClientStateAction.ConnectionLost, ClientTransportState.Closed)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.OpenStart, ClientTransportState.Opening)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.OpenStart, ClientTransportState.Open)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.OpenStart, ClientTransportState.Closing)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.OpenStart, ClientTransportState.Closed)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.OpenSuccess, ClientTransportState.Opening)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.OpenSuccess, ClientTransportState.Open)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.OpenSuccess, ClientTransportState.Closing)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.OpenSuccess, ClientTransportState.Closed)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.OpenFailure, ClientTransportState.Opening)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.OpenFailure, ClientTransportState.Open)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.OpenFailure, ClientTransportState.Closing)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.OpenFailure, ClientTransportState.Closed)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.CloseStart, ClientTransportState.Opening)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.CloseStart, ClientTransportState.Open)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.CloseStart, ClientTransportState.Closing)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.CloseStart, ClientTransportState.Closed)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.CloseComplete, ClientTransportState.Opening)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.CloseComplete, ClientTransportState.Open)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.CloseComplete, ClientTransportState.Closing)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.ConnectionLost, ClientTransportState.Opening)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.ConnectionLost, ClientTransportState.Open)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.ConnectionLost, ClientTransportState.Closing)]
+        [DataRow(ClientTransportState.Closing, ClientStateAction.ConnectionLost, ClientTransportState.Closed)]
+        public void ClientTransportStateMachine_InvalidStateTransition(object initialState, object action, object nextState)
```

Comment:
These are all of the invalid state transitions.

1. Let me know if you can think of a better way to test this.
2. Let me know if a state transition listed here looks valid to you.

Reply:
Aside from breaking this up into three categories, `startFromOpen`, `startFromClosing` and `startFromClosed` or something like that, I can't think of another way to make this read better.

Reply:
I like that idea. I'll split these in a different PR.